File: src/utils/model-metadata.ts
import * as fs from "fs";
import * as path from "path";
import { Logger, getLogger } from "./logger";

/**
 * Interface for discrete action space configuration
 */
export interface DiscreteAction {
  steering_angle: number;
  speed: number;
}

/**
 * Interface for continuous action space range
 */
export interface ContinuousRange {
  low: number;
  high: number;
}

/**
 * Interface for continuous action space configuration
 */
export interface ContinuousActionSpace {
  speed: ContinuousRange;
  steering_angle: ContinuousRange;
}

/**
 * Type definition for action space which can be either discrete or continuous
 */
export type ActionSpace = DiscreteAction[] | ContinuousActionSpace;

/**
 * Action space type enumeration
 */
export enum ActionSpaceType {
  DISCRETE = "discrete",
  CONTINUOUS = "continuous",
}

/**
 * Neural network type enumeration
 */
export enum NeuralNetworkType {
  DEEP_CONVOLUTIONAL_NETWORK_SHALLOW = "DEEP_CONVOLUTIONAL_NETWORK_SHALLOW",
  DEEP_CONVOLUTIONAL_NETWORK = "DEEP_CONVOLUTIONAL_NETWORK",
  LLM = "LLM",
}

/**
 * Training algorithm type enumeration
 */
export enum TrainingAlgorithm {
  CLIPPED_PPO = "clipped_ppo",
  SAC = "sac",
}

/**
 * Sensor type enumeration
 */
export enum SensorType {
  FRONT_FACING_CAMERA = "FRONT_FACING_CAMERA",
  STEREO_CAMERAS = "STEREO_CAMERAS",
  LIDAR = "LIDAR",
}

/**
 * Interface for LLM configuration
 */
export interface LLMConfig {
  model_id: string;
  max_tokens: number;
  system_prompt: string;
  repeated_prompt: string;
  context_window: number;
}

/**
 * Interface for model metadata configuration
 */
export interface ModelMetadata {
  action_space: ActionSpace;
  sensor: SensorType[];
  neural_network: NeuralNetworkType;
  training_algorithm?: TrainingAlgorithm;
  action_space_type: ActionSpaceType;
  version: string;
  llm_config?: LLMConfig;
}

/**
 * Class to handle loading and processing model metadata
 */
export class ModelMetadataHandler {
  private logger: Logger;
  private metadata: ModelMetadata | null = null;

  constructor() {
    this.logger = getLogger("ModelMetadata");
  }

  /**
   * Load model metadata from the specified JSON file
   *
   * @param filePath Path to the model metadata JSON file
   * @returns Promise resolving to the loaded model metadata
   */
  public loadModelMetadata(
    filePath: string = "model_metadata.json"
  ): ModelMetadata {
    try {
      // Resolve path relative to project root
      const resolvedPath = path.resolve(process.cwd(), filePath);

      this.logger.debug(`Loading model metadata from: ${resolvedPath}`);

      if (!fs.existsSync(resolvedPath)) {
        throw new Error(`Model metadata file not found: ${resolvedPath}`);
      }

      const fileContent = fs.readFileSync(resolvedPath, "utf8");
      this.metadata = JSON.parse(fileContent) as ModelMetadata;

      // Validate the loaded metadata
      this.validateMetadata();

      this.logger.info(
        `Loaded model metadata with ${this.getActionSpaceType()} action space and ${
          this.metadata.neural_network
        } neural network type`
      );

      return this.metadata;
    } catch (error) {
      this.logger.error(`Failed to load model metadata: ${error}`);
      throw error;
    }
  }

  /**
   * Get the current model metadata
   * @returns The model metadata or null if not loaded
   */
  public getMetadata(): ModelMetadata | null {
    return this.metadata;
  }

  /**
   * Validate the loaded metadata structure
   */
  private validateMetadata(): void {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    // Check required fields
    if (!this.metadata.action_space) {
      throw new Error("Missing action_space in model metadata");
    }

    if (!this.metadata.sensor || !Array.isArray(this.metadata.sensor)) {
      throw new Error(
        "Missing or invalid sensor configuration in model metadata"
      );
    }

    if (!this.metadata.neural_network) {
      throw new Error("Missing neural_network in model metadata");
    }

    // Check for LLM configuration when neural network is LLM
    if (this.metadata.neural_network === NeuralNetworkType.LLM) {
      if (!this.metadata.llm_config) {
        throw new Error(
          "LLM neural network type requires llm_config to be specified"
        );
      }

      const llmConfig = this.metadata.llm_config;
      if (!llmConfig.model_id) {
        throw new Error("Missing model_id in llm_config");
      }

      if (typeof llmConfig.max_tokens !== "number") {
        throw new Error("Missing or invalid max_tokens in llm_config");
      }

      if (typeof llmConfig.system_prompt !== "string") {
        throw new Error("Missing system_prompt in llm_config");
      }

      if (typeof llmConfig.repeated_prompt !== "string") {
        throw new Error("Missing repeated_prompt in llm_config");
      }

      if (
        typeof llmConfig.context_window !== "number" ||
        llmConfig.context_window < 0
      ) {
        throw new Error("Invalid context_window in llm_config");
      }
    } else {
      // Traditional neural networks need training algorithm
      if (!this.metadata.training_algorithm) {
        throw new Error(
          "Missing training_algorithm in model metadata for traditional neural network"
        );
      }
    }

    // Validate action space based on type
    const actionSpaceType = this.getActionSpaceType();

    if (actionSpaceType === ActionSpaceType.DISCRETE) {
      // For discrete, action_space should be an array of actions
      if (!Array.isArray(this.metadata.action_space)) {
        throw new Error("Discrete action space should be an array");
      }

      // Validate each action
      for (const action of this.metadata.action_space as DiscreteAction[]) {
        if (
          typeof action.steering_angle !== "number" ||
          typeof action.speed !== "number"
        ) {
          throw new Error("Invalid discrete action format");
        }
      }
    } else if (actionSpaceType === ActionSpaceType.CONTINUOUS) {
      // For continuous, action_space should be an object with ranges
      const continuousSpace = this.metadata
        .action_space as ContinuousActionSpace;

      if (!continuousSpace.speed || !continuousSpace.steering_angle) {
        throw new Error("Continuous action space missing required ranges");
      }

      if (
        typeof continuousSpace.speed.low !== "number" ||
        typeof continuousSpace.speed.high !== "number" ||
        typeof continuousSpace.steering_angle.low !== "number" ||
        typeof continuousSpace.steering_angle.high !== "number"
      ) {
        throw new Error("Invalid continuous action range format");
      }

      // Validate ranges
      if (continuousSpace.speed.low >= continuousSpace.speed.high) {
        throw new Error("Speed range low must be less than high");
      }

      if (
        continuousSpace.steering_angle.low >=
        continuousSpace.steering_angle.high
      ) {
        throw new Error("Steering angle range low must be less than high");
      }
    }
  }

  /**
   * Get the action space type from the metadata
   * @returns ActionSpaceType enum value
   */
  public getActionSpaceType(): ActionSpaceType {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    return (
      this.metadata.action_space_type ||
      (Array.isArray(this.metadata.action_space)
        ? ActionSpaceType.DISCRETE
        : ActionSpaceType.CONTINUOUS)
    );
  }

  /**
   * Get the action space configuration
   * @returns Action space configuration
   */
  public getActionSpace(): ActionSpace {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    return this.metadata.action_space;
  }

  /**
   * Check if the model uses an LLM
   * @returns True if the model uses an LLM, false otherwise
   */
  public isLLMModel(): boolean {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    return this.metadata.neural_network === NeuralNetworkType.LLM;
  }

  /**
   * Get the LLM configuration
   * @returns LLM configuration or null if not an LLM model
   */
  public getLLMConfig(): LLMConfig | null {
    if (!this.metadata || !this.isLLMModel()) {
      return null;
    }

    return this.metadata.llm_config || null;
  }

  /**
   * Get discrete actions from metadata
   * @returns Array of discrete actions or null if action space is not discrete
   */
  public getDiscreteActions(): DiscreteAction[] | null {
    if (
      !this.metadata ||
      this.getActionSpaceType() !== ActionSpaceType.DISCRETE
    ) {
      return null;
    }

    return this.metadata.action_space as DiscreteAction[];
  }

  /**
   * Get continuous action space ranges
   * @returns Continuous action space ranges or null if action space is not continuous
   */
  public getContinuousActionSpace(): ContinuousActionSpace | null {
    if (
      !this.metadata ||
      this.getActionSpaceType() !== ActionSpaceType.CONTINUOUS
    ) {
      return null;
    }

    return this.metadata.action_space as ContinuousActionSpace;
  }

  /**
   * Check if a steering angle is within the valid range
   * @param steeringAngle The steering angle to check
   * @returns Whether the steering angle is valid
   */
  public isValidSteeringAngle(steeringAngle: number): boolean {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    if (this.getActionSpaceType() === ActionSpaceType.CONTINUOUS) {
      const { steering_angle } = this.metadata
        .action_space as ContinuousActionSpace;
      return (
        steeringAngle >= steering_angle.low &&
        steeringAngle <= steering_angle.high
      );
    } else {
      const actions = this.metadata.action_space as DiscreteAction[];
      return actions.some((action) => action.steering_angle === steeringAngle);
    }
  }

  /**
   * Check if a speed value is within the valid range
   * @param speed The speed to check
   * @returns Whether the speed is valid
   */
  public isValidSpeed(speed: number): boolean {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    if (this.getActionSpaceType() === ActionSpaceType.CONTINUOUS) {
      const { speed: speedRange } = this.metadata
        .action_space as ContinuousActionSpace;
      return speed >= speedRange.low && speed <= speedRange.high;
    } else {
      const actions = this.metadata.action_space as DiscreteAction[];
      return actions.some((action) => action.speed === speed);
    }
  }

  /**
   * Get the neural network type from metadata
   * @returns Neural network type
   */
  public getNeuralNetworkType(): NeuralNetworkType {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    return this.metadata.neural_network;
  }

  /**
   * Get the training algorithm from metadata
   * @returns Training algorithm or undefined if using LLM
   */
  public getTrainingAlgorithm(): TrainingAlgorithm | undefined {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    return this.metadata.training_algorithm;
  }

  /**
   * Get the sensors configuration from metadata
   * @returns Array of sensors
   */
  public getSensors(): SensorType[] {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    return this.metadata.sensor;
  }

  /**
   * Find the closest discrete action to the given continuous values
   * Only applicable for discrete action spaces
   *
   * @param steeringAngle Target steering angle
   * @param speed Target speed
   * @returns The closest discrete action
   */
  public findClosestDiscreteAction(
    steeringAngle: number,
    speed: number
  ): DiscreteAction | null {
    if (
      !this.metadata ||
      this.getActionSpaceType() !== ActionSpaceType.DISCRETE
    ) {
      return null;
    }

    const actions = this.metadata.action_space as DiscreteAction[];

    let closestAction = actions[0];
    let minDistance = Infinity;

    for (const action of actions) {
      // Calculate Euclidean distance in the action space
      const steeringDiff = action.steering_angle - steeringAngle;
      const speedDiff = action.speed - speed;
      const distance = Math.sqrt(
        steeringDiff * steeringDiff + speedDiff * speedDiff
      );

      if (distance < minDistance) {
        minDistance = distance;
        closestAction = action;
      }
    }

    return closestAction;
  }

  /**
   * Convert a continuous action to a valid discrete action if needed
   *
   * @param steeringAngle Steering angle value
   * @param speed Speed value
   * @returns Valid action based on the action space type
   */
  public normalizeAction(steeringAngle: number, speed: number): DiscreteAction {
    if (!this.metadata) {
      throw new Error("No metadata loaded");
    }

    if (this.getActionSpaceType() === ActionSpaceType.CONTINUOUS) {
      const continuousSpace = this.metadata
        .action_space as ContinuousActionSpace;

      // Clamp values to the valid ranges
      const normalizedSteeringAngle = Math.max(
        continuousSpace.steering_angle.low,
        Math.min(continuousSpace.steering_angle.high, steeringAngle)
      );

      const normalizedSpeed = Math.max(
        continuousSpace.speed.low,
        Math.min(continuousSpace.speed.high, speed)
      );

      return {
        steering_angle: normalizedSteeringAngle,
        speed: normalizedSpeed,
      };
    } else {
      // Find closest discrete action
      const closestAction = this.findClosestDiscreteAction(
        steeringAngle,
        speed
      );
      if (!closestAction) {
        throw new Error("Failed to find a valid discrete action");
      }
      return closestAction;
    }
  }
}

// Export a singleton instance for convenience
export const modelMetadata = new ModelMetadataHandler();

// Default export for backwards compatibility
export default modelMetadata;

File: deepracer_llm_agent/utils/model_metadata.py
import json
import os
import logging
import math
from enum import Enum
from typing import Dict, List, Union, Optional, TypedDict, Any, Tuple

# Configure logger
logger = logging.getLogger("ModelMetadata")

# Enums that mirror the TypeScript implementation
class ActionSpaceType(str, Enum):
    DISCRETE = "discrete"
    CONTINUOUS = "continuous"

class NeuralNetworkType(str, Enum):
    DEEP_CONVOLUTIONAL_NETWORK_SHALLOW = "DEEP_CONVOLUTIONAL_NETWORK_SHALLOW"
    DEEP_CONVOLUTIONAL_NETWORK = "DEEP_CONVOLUTIONAL_NETWORK"
    LLM = "LLM"

class TrainingAlgorithm(str, Enum):
    CLIPPED_PPO = "clipped_ppo"
    SAC = "sac"

class SensorType(str, Enum):
    FRONT_FACING_CAMERA = "FRONT_FACING_CAMERA"
    STEREO_CAMERAS = "STEREO_CAMERAS"
    LIDAR = "LIDAR"

# Type definitions for action spaces and metadata
class DiscreteAction(TypedDict):
    steering_angle: float
    speed: float

class ContinuousRange(TypedDict):
    low: float
    high: float

class ContinuousActionSpace(TypedDict):
    speed: ContinuousRange
    steering_angle: ContinuousRange

# Define ActionSpace as a Union type
ActionSpace = Union[List[DiscreteAction], ContinuousActionSpace]

class LLMConfig(TypedDict):
    model_id: str
    max_tokens: int
    system_prompt: Union[str, List[str]]
    repeated_prompt: str
    context_window: int

# Model metadata type
class ModelMetadata(TypedDict, total=False):
    action_space: ActionSpace
    sensor: List[SensorType]
    neural_network: NeuralNetworkType
    training_algorithm: Optional[TrainingAlgorithm]
    action_space_type: ActionSpaceType
    version: str
    llm_config: Optional[LLMConfig]

class ModelMetadataHandler:
    """Class to handle loading and processing model metadata"""
    
    def __init__(self):
        self.metadata = None
    
    def load_model_metadata(self, file_path: str = "model_metadata.json") -> ModelMetadata:
        """
        Load model metadata from the specified JSON file
        
        Args:
            file_path: Path to the model metadata JSON file
            
        Returns:
            The loaded model metadata
            
        Raises:
            FileNotFoundError: If the model metadata file is not found
            json.JSONDecodeError: If the file contains invalid JSON
            ValueError: If the metadata validation fails
        """
        try:
            # Resolve path relative to current working directory
            resolved_path = os.path.abspath(file_path)
            logger.debug(f"Loading model metadata from: {resolved_path}")
            
            if not os.path.exists(resolved_path):
                raise FileNotFoundError(f"Model metadata file not found: {resolved_path}")
            
            with open(resolved_path, 'r') as file:
                self.metadata = json.load(file)
            
            # Validate the loaded metadata
            self._validate_metadata()
            
            logger.debug(
                f"Loaded model metadata with {self.get_action_space_type()} action space and "
                f"{self.metadata['neural_network']} neural network type"
            )
            
            return self.metadata
        
        except Exception as error:
            logger.error(f"Failed to load model metadata: {error}")
            raise
    
    def get_metadata(self) -> Optional[ModelMetadata]:
        """
        Get the current model metadata
        
        Returns:
            The model metadata or None if not loaded
        """
        return self.metadata
    
    def _validate_metadata(self) -> None:
        """
        Validate the loaded metadata structure
        
        Raises:
            ValueError: If the metadata is invalid
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        # Check required fields
        if "action_space" not in self.metadata:
            raise ValueError("Missing action_space in model metadata")
        
        if "sensor" not in self.metadata or not isinstance(self.metadata["sensor"], list):
            raise ValueError("Missing or invalid sensor configuration in model metadata")
        
        if "neural_network" not in self.metadata:
            raise ValueError("Missing neural_network in model metadata")
        
        # Check for LLM configuration when neural network is LLM
        if self.metadata["neural_network"] == NeuralNetworkType.LLM.value:
            if "llm_config" not in self.metadata:
                raise ValueError("LLM neural network type requires llm_config to be specified")
            
            llm_config = self.metadata["llm_config"]
            if "model_id" not in llm_config:
                raise ValueError("Missing model_id in llm_config")
            
            if not isinstance(llm_config.get("max_tokens"), int):
                raise ValueError("Missing or invalid max_tokens in llm_config")
            
            if not isinstance(llm_config.get("system_prompt"), (str, list)):
                raise ValueError("Missing or invalid system_prompt in llm_config")
            
            if not isinstance(llm_config.get("context_window"), int) or llm_config.get("context_window", 0) < 0:
                raise ValueError("Invalid context_window in llm_config")
        else:
            # Traditional neural networks need training algorithm
            if "training_algorithm" not in self.metadata:
                raise ValueError("Missing training_algorithm for traditional neural network")
        
        # Validate action space based on type
        action_space_type = self.get_action_space_type()
        
        if action_space_type == ActionSpaceType.DISCRETE:
            # For discrete, action_space should be a list of actions
            if not isinstance(self.metadata["action_space"], list):
                raise ValueError("Discrete action space should be a list")
            
            # Validate each action
            for action in self.metadata["action_space"]:
                if not isinstance(action.get("steering_angle"), (int, float)) or \
                   not isinstance(action.get("speed"), (int, float)):
                    raise ValueError("Invalid discrete action format")
        
        elif action_space_type == ActionSpaceType.CONTINUOUS:
            # For continuous, action_space should be an object with ranges
            continuous_space = self.metadata["action_space"]
            
            if "speed" not in continuous_space or "steering_angle" not in continuous_space:
                raise ValueError("Continuous action space missing required ranges")
            
            if not isinstance(continuous_space["speed"].get("low"), (int, float)) or \
               not isinstance(continuous_space["speed"].get("high"), (int, float)) or \
               not isinstance(continuous_space["steering_angle"].get("low"), (int, float)) or \
               not isinstance(continuous_space["steering_angle"].get("high"), (int, float)):
                raise ValueError("Invalid continuous action range format")
            
            # Validate ranges
            if continuous_space["speed"]["low"] >= continuous_space["speed"]["high"]:
                raise ValueError("Speed range low must be less than high")
            
            if continuous_space["steering_angle"]["low"] >= continuous_space["steering_angle"]["high"]:
                raise ValueError("Steering angle range low must be less than high")
    
    def get_action_space_type(self) -> ActionSpaceType:
        """
        Get the action space type from the metadata
        
        Returns:
            ActionSpaceType enum value
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        # Check if action_space_type is explicitly specified
        if "action_space_type" in self.metadata:
            return ActionSpaceType(self.metadata["action_space_type"])
        
        # Otherwise infer from action_space structure
        return (ActionSpaceType.DISCRETE if isinstance(self.metadata["action_space"], list)
                else ActionSpaceType.CONTINUOUS)
    
    def get_action_space(self) -> ActionSpace:
        """
        Get the action space configuration
        
        Returns:
            Action space configuration
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        return self.metadata["action_space"]
    
    def is_llm_model(self) -> bool:
        """
        Check if the model uses an LLM
        
        Returns:
            True if the model uses an LLM, False otherwise
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        return self.metadata["neural_network"] == NeuralNetworkType.LLM.value
    
    def get_llm_config(self) -> Optional[LLMConfig]:
        """
        Get the LLM configuration
        
        Returns:
            LLM configuration or None if not an LLM model
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata or not self.is_llm_model():
            return None
        
        return self.metadata.get("llm_config")
    
    def get_discrete_actions(self) -> Optional[List[DiscreteAction]]:
        """
        Get discrete actions from metadata
        
        Returns:
            Array of discrete actions or None if action space is not discrete
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata or self.get_action_space_type() != ActionSpaceType.DISCRETE:
            return None
        
        return self.metadata["action_space"]
    
    def get_continuous_action_space(self) -> Optional[ContinuousActionSpace]:
        """
        Get continuous action space ranges
        
        Returns:
            Continuous action space ranges or None if action space is not continuous
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata or self.get_action_space_type() != ActionSpaceType.CONTINUOUS:
            return None
        
        return self.metadata["action_space"]
    
    def is_valid_steering_angle(self, steering_angle: float) -> bool:
        """
        Check if a steering angle is within the valid range
        
        Args:
            steering_angle: The steering angle to check
            
        Returns:
            Whether the steering angle is valid
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        if self.get_action_space_type() == ActionSpaceType.CONTINUOUS:
            steering_range = self.metadata["action_space"]["steering_angle"]
            return steering_angle >= steering_range["low"] and steering_angle <= steering_range["high"]
        else:
            actions = self.metadata["action_space"]
            return any(action["steering_angle"] == steering_angle for action in actions)
    
    def is_valid_speed(self, speed: float) -> bool:
        """
        Check if a speed value is within the valid range
        
        Args:
            speed: The speed to check
            
        Returns:
            Whether the speed is valid
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        if self.get_action_space_type() == ActionSpaceType.CONTINUOUS:
            speed_range = self.metadata["action_space"]["speed"]
            return speed >= speed_range["low"] and speed <= speed_range["high"]
        else:
            actions = self.metadata["action_space"]
            return any(action["speed"] == speed for action in actions)
    
    def get_neural_network_type(self) -> NeuralNetworkType:
        """
        Get the neural network type from metadata
        
        Returns:
            Neural network type
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        return NeuralNetworkType(self.metadata["neural_network"])
    
    def get_training_algorithm(self) -> Optional[TrainingAlgorithm]:
        """
        Get the training algorithm from metadata
        
        Returns:
            Training algorithm or None if using LLM
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        if "training_algorithm" not in self.metadata:
            return None
            
        return TrainingAlgorithm(self.metadata["training_algorithm"])
    
    def get_sensors(self) -> List[SensorType]:
        """
        Get the sensors configuration from metadata
        
        Returns:
            List of sensors
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        return [SensorType(sensor) for sensor in self.metadata["sensor"]]
    
    def find_closest_discrete_action(self, steering_angle: float, speed: float) -> Optional[DiscreteAction]:
        """
        Find the closest discrete action to the given continuous values
        Only applicable for discrete action spaces
        
        Args:
            steering_angle: Target steering angle
            speed: Target speed
            
        Returns:
            The closest discrete action or None if not applicable
        """
        if not self.metadata or self.get_action_space_type() != ActionSpaceType.DISCRETE:
            return None
        
        actions = self.metadata["action_space"]
        
        closest_action = actions[0]
        min_distance = float('inf')
        
        for action in actions:
            # Calculate Euclidean distance in the action space
            steering_diff = action["steering_angle"] - steering_angle
            speed_diff = action["speed"] - speed
            distance = math.sqrt(steering_diff * steering_diff + speed_diff * speed_diff)
            
            if distance < min_distance:
                min_distance = distance
                closest_action = action
        
        return closest_action
    
    def normalize_action(self, steering_angle: float, speed: float) -> DiscreteAction:
        """
        Convert a continuous action to a valid discrete action if needed
        
        Args:
            steering_angle: Steering angle value
            speed: Speed value
            
        Returns:
            Valid action based on the action space type
            
        Raises:
            ValueError: If no metadata is loaded
        """
        if not self.metadata:
            raise ValueError("No metadata loaded")
        
        if self.get_action_space_type() == ActionSpaceType.CONTINUOUS:
            continuous_space = self.metadata["action_space"]
            
            # Check if values exceed the valid ranges and log warnings
            if (steering_angle < continuous_space["steering_angle"]["low"] or 
                steering_angle > continuous_space["steering_angle"]["high"]):
                logger.warning(
                    f"Steering angle {steering_angle} exceeds valid range "
                    f"[{continuous_space['steering_angle']['low']}, {continuous_space['steering_angle']['high']}]"
                )
            
            if (speed < continuous_space["speed"]["low"] or 
                speed > continuous_space["speed"]["high"]):
                logger.warning(
                    f"Speed {speed} exceeds valid range "
                    f"[{continuous_space['speed']['low']}, {continuous_space['speed']['high']}]"
                )
            
            # Clamp values to the valid ranges
            normalized_steering_angle = max(
                continuous_space["steering_angle"]["low"],
                min(continuous_space["steering_angle"]["high"], steering_angle)
            )
            
            normalized_speed = max(
                continuous_space["speed"]["low"],
                min(continuous_space["speed"]["high"], speed)
            )
            
            return {
                "steering_angle": normalized_steering_angle,
                "speed": normalized_speed
            }
        else:
            # For discrete action spaces, log a warning if the requested action is not exact
            discrete_actions = self.metadata["action_space"]
            exact_match = any(
                action["steering_angle"] == steering_angle and action["speed"] == speed
                for action in discrete_actions
            )
            
            if not exact_match:
                logger.warning(
                    f"Requested action ({steering_angle}, {speed}) is not in the discrete action space, "
                    f"finding closest match"
                )
            
            # Find closest discrete action
            closest_action = self.find_closest_discrete_action(steering_angle, speed)
            if not closest_action:
                raise ValueError("Failed to find a valid discrete action")
            
            return closest_action


# Create a singleton instance for convenience
model_metadata = ModelMetadataHandler()
